src/worker/mod.rs

```diff
@@ -337,15 +375,20 @@ impl<
         }))
     }
 
-    /// Run this worker on the given `queues` until an I/O error occurs (`Err` is returned), or
-    /// until the server tells the worker to disengage (`Ok` is returned).
+    /// Run this worker on the given `queues`.
+    ///
+    /// Will run the worker until an I/O error occurs (`Err` is returned), or until the server tells the worker
+    /// to disengage (`Ok` is returned), or a signal from the user-space code has been received via a future
+    /// supplied to [`WorkerBuilder::with_graceful_shutdown`](`Ok` is returned).
     ///
-    /// The value in an `Ok` indicates the number of workers that may still be processing jobs.
+    /// The value in an `Ok` holds a tuple with the reason why the run has discontinued (see [`RunCeaseReason`])
+    /// and the number of workers that may still be processing jobs. Note that `0` can also indicate that
+    /// the [graceful shutdown period](WorkerBuilder::shutdown_timeout) has been exceeded.
     ///
     /// If an error occurred while reporting a job success or failure, the result will be re-reported to the server
     /// without re-executing the job. If the worker was terminated (i.e., `run` returns  with an `Ok` response),
     /// the worker should **not** try to resume by calling `run` again. This will cause a panic.
-    pub async fn run<Q>(&mut self, queues: &[Q]) -> Result<usize, Error>
+    pub async fn run<Q>(&mut self, queues: &[Q]) -> Result<(RunCeaseReason, usize), Error>
```

Comment:
How do you feel about returning a `struct` here with `pub` fields rather than a tuple?

Reply:
Introduced a `StopDetails` that holds the reason and nrunning and put it to a dedicated module.
Do you think we should  keep this stop details container non_exhaustive as well?

And yes - `run cease` is  really nothing but `stop`. :) The idea was to prefix this guy with the method name, i.e. run - run-cease.

Reply:
Yeah, I think that should probably also be `#[non_exhaustive]`. The main downside to doing so is that people can't easily construct their own for testing purposes, but if we have it also `derive(Default)` then maybe that helps with that?